api/v1/openapi.yaml

```diff
@@ -824,8 +824,13 @@ definitions:
         description: The policy revision this endpoint is running on
         type: integer
       proxy-policy-revision:
-        description: The policy revision currently enforced in the proxy
+        description: The policy revision currently enforced in the proxy for this endpoint
         type: integer
+      proxy-statistics:
+        description: Statistics of the proxy redirects configured for this endpoint
+        type: array
+        items:
+          "$ref": "#/definitions/ProxyRedirectStatistics"
```

Comment:
This split seems weird, why not just encode the entire proxy redirect status in the endpoint? Splitting this means a user has to manually correlate the statistics via the endpoint and rules and proxy port via the global status.

Reply:
Because we won't have that information from sidecar proxies.

Reply:
Kafka shouldn't break.

Reply:
One advantage of aggregating at endpoint level is that we'll have stats that will span the whole lifetime of the endpoint, not just the lifetime of redirects. I think that's more useful.

Reply:
And we can't correlate stats from sidecar proxies with redirects, because we don't even configure redirects in this case. That's the whole point of this PR.

Reply:
I agree that having the stats in the endpoint is neat. I also understand that there is no conventional proxy port, although it would be great to see the port number on which Envoy is listening. My main concern left is that in order to see the proxy rules, the user has to correlate the entry in the status.

Users will also get used to seeing the proxy status in either place and it will just be empty for the sidecar model. A user will not want to understand what a redirect and why it is missing for the sidecar case. Can't we move everything to the endpoint and just not call it a redirect?

Reply:
We can't get the redirect port from sidecars we don't manage, e.g. Istio sidecars. And even if we did it wouldn't be useful information. That port is 15000 by default for every redirect across the whole cluster.

I could add the redirect port into the Endpoint's `proxy-statistics` if there's a redirect currently active for the port and if the redirect port is known:
```
 "proxy-statistics": [
          {
            "location": "ingress",
            "port": 80,
            "protocol": "http",
            "redirect-port": 12345,
            "statistics": {
              ...
            }
          }
        ],
```

And if there's currently no redirect for the port, or the redirect port is not known, there will be no `"redirect-port"` field:
```
 "proxy-statistics": [
          {
            "location": "ingress",
            "port": 80,
            "protocol": "http",
            "statistics": {
              ...
            }
          }
        ],
```

And then I can remove the redirects from `cilium status` altogether.

The only drawback is that it will be hard to know at a glance what redirects are currently installed into the Envoy and Kafka proxies managed by Cilium (not in sidecars).

@tgraf, would you be OK with that API?

Reply:
And regarding the rules enforced in the proxy, now it’s even neater because with NPDS we now push the endpoints’ full policies to the Envoy proxies. And an endpoint’s policy is right in the same object returned by `cilium endpoint get`.

When we move the Kafka proxy to use xDS / NPDS for configuration, that will also be true for Kafka. We will NOT configure rules specifically for each redirect.

Reply:
> That port is 15000 by default for every redirect across the whole cluster.

It may be clear to you, it will not be clear to users and it will be useful information to the user as there will be traffic inside of the pod to and from that port. As long as we can add it later on, I'm fine.

> And regarding the rules enforced in the proxy, now it’s even neater because with NPDS we now push the endpoints’ full policies to the Envoy proxies. And an endpoint’s policy is right in the same object returned by cilium endpoint get.

That's great, given we are just before 1.0, we absolutely don't want to expose something in the overall status and then remove it. How difficult is it to just expose the kafka rules like this as well? I really dislike exposing any information in different places depending on whether the proxy model is sidecar or not.

Reply:
We'll have to move the Kafka proxy to use NPDS when we move it into a separate process. When that's done, the policy that is returned in `cilium endpoint get` will be the one that is pushed to the Kafka proxy.

I'll remove `redirects` altogether from the proxy status then.